packages/oidc-dpop-client-browser/src/dpop/tokenExchange.ts

```diff
@@ -0,0 +1,233 @@
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+ * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+import { IClient, IIssuerConfig } from "../common/types";
+import { JSONWebKey } from "jose";
+import { createDpopHeader, decodeJwt } from "./dpop";
+import { generateJwkForDpop } from "./keyGen";
+import formurlencoded from "form-urlencoded";
+
+function hasAccessToken(
+  value: { access_token: string } | Record<string, unknown>
+): value is { access_token: string } {
+  return value.access_token && typeof value.access_token === "string";
+}
+
+function hasIdToken(
+  value: { id_token: string } | Record<string, unknown>
+): value is { id_token: string } {
+  return value.id_token && typeof value.id_token === "string";
+}
+
+function hasRefreshToken(
+  value: { refresh_token: string } | Record<string, unknown>
+): value is { refresh_token: string } {
+  return value.refresh_token && typeof value.refresh_token === "string";
+}
+
+function hasTokenType(
+  value: { token_type: string } | Record<string, unknown>
+): value is { token_type: string } {
+  return value.token_type && typeof value.token_type === "string";
+}
+
+export type TokenEndpointResponse = {
+  accessToken: string;
+  idToken: string;
+  webid: string;
+  refreshToken?: string;
+  dpopJwk?: string;
+};
+
+export type TokenEndpointInput = {
+  grantType: string;
+  redirectUri: string;
+  code: string;
+  codeVerifier: string;
+};
+
+type WebidOidcIdToken = {
+  sub: string;
+  iss: string;
+  webid?: string;
+};
+
+function isWebidOidcIdToken(
+  token: WebidOidcIdToken | Record<string, unknown>
+): token is WebidOidcIdToken {
+  return (
+    (token.sub &&
+      typeof token.sub === "string" &&
+      token.iss &&
+      typeof token.iss === "string" &&
+      !token.webid) ||
+    typeof token.webid === "string"
+  );
+}
+
+/**
+ * Extracts a Webid from an ID token based on https://github.com/solid/webid-oidc-spec.
+ * The upcoming spec is still a work in progress.
+ *
+ * Note: this function does not implement the userinfo webid lookup yet.
+ * @param idToken
+ */
+async function deriveWebidFromIdToken(idToken: string): Promise<string> {
+  const decoded = await decodeJwt(idToken);
+  if (!isWebidOidcIdToken(decoded)) {
+    throw new Error(
+      `Invalid ID token: ${JSON.stringify(
+        decoded
+      )} is missing 'sub' or 'iss' claims`
+    );
+  }
+  if (decoded.webid) {
+    return decoded.webid;
+  }
+  if (!decoded.sub.match(/^https?:\/\/.+\..+$/)) {
+    throw new Error(
+      `Cannot extract WebID from ID token: the ID token returned by ${decoded.iss} has no webid claim, nor an IRI-like sub claim: [${decoded.sub}]`
```

Comment:
Suggest: `Cannot extract WebID from ID token: the ID token returned by [${decoded.iss}] has no 'webid' claim, nor an IRI-like 'sub' claim: [${decoded.sub}]`